src/MSMToolsCluster.jl

```diff
@@ -0,0 +1,259 @@
+using LinearAlgebra
+using Arpack
+using Statistics
+using SparseArrays
+using LightGraphs
+using Optim
+
+# NOTE: Reimplementation in Julia for select functions from MSM Tools (referenced in DeepTime)
+# https://github.com/markovmodel/msmtools
+# https://github.com/deeptime-ml/deeptime/tree/a6ac0b93a55d688fe8f3af119680105763366220/deeptime/markov/tools
+
+# Reimplementation of PCCA+ spectral clustering method with optimized memberships
+# https://github.com/deeptime-ml/deeptime/blob/a6ac0b93a55d688fe8f3af119680105763366220/deeptime/markov/tools/analysis/dense/_pcca.py#L288
+function pcca_memberships(P::Matrix{Float64}, m::Int, π::Vector{Float64}=nothing)
+    n = size(P, 1)
+    if m > n
+        throw(ArgumentError("Number of metastable states m=$m exceeds number of states of transition matrix n=$n"))
+    end
+    println("P: $P")
+    if !is_transition_matrix(P)
+        throw(ArgumentError("Input matrix is not a transition matrix."))
+    end
+    if isnothing(π)
+        if ndims(π) > 1
+            throw(ArgumentError("Expected 1D stationary distribution array"))
+        elseif length(π) != n
+            throw(ArgumentError("Stationary distribution must be defined on entire space, piecewise if the transition matrix " *
+            "has multiple connected components. It covered $(length(π)) != $n states."))
+        end
+    end
+
+    chi = zeros(n, m)
+    components = connected_sets(P)
+    closed_components = []
+    transition_states = []
+    for component in components
+        rest = setdiff(1:n, component)
+        # if component is closed, store with positive equilibrium distribution
+        if sum(P[component, rest]) == 0
+            push!(closed_components, component)
+        # otherwise store as transition states with vanishing equilibrium distribution
+        else
+            push!(transition_states, component)
+        end
+    end
+
+    n_closed_components = length(closed_components)
+    closed_states = vcat(closed_components...)
```

Comment:
Using `vcat` is a very good idea. Often it's better to use `reduce(vcat, components)` instead of `vcat(components...)` to avoid an allocation

Reply:
fixed in 3c9f26d25c8004d3466bdb9da47d155bf213a32e